modules/free/src/main/scala/doobie/free/kleisliinterpreter.scala

```diff
@@ -100,11 +101,18 @@ class KleisliInterpreter[M[_]](logHandler: LogHandler[M])(implicit val asyncM: W
   def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli { a =>
     // primitive JDBC methods throw exceptions and so do we when reading values
     // so catch any non-fatal exceptions and lift them into the effect
-    try {
-      asyncM.blocking(f(a))
-    } catch {
-      case scala.util.control.NonFatal(e) => asyncM.raiseError(e)
-    }
+    import cats.syntax.all._
+    for {
+      jdbcBlockedFiber <- asyncM.start(asyncM.blocking(f(a)).attempt)
+      a <- jdbcBlockedFiber.join.flatMap {
```

Comment:
I've got some concerns regarding this code. They may or may not be valid though, but let me pronounce them:
- `Sync#blocking` is non-cancellable by design. Therefore the execution pipeline can be cancelled either before entering `blocking` or right after it completes, but not inside. Therefore it can only cancel a fiber that runs the query, but not the operation itself.
- There can be a potential fiber leak with this code: if a cancel signal comes when the fiber is created, but before it gets joined, then here it goes (<sub>btw, I believe this is one of the reasons why `background` is almost always preferable over low-level fiber start/join manipulations</sub>).
- Firing a fiber and then joining it immediately after looks quite suspicious. What is it attempted to accomplish there exactly?

Reply:
> Therefore it can only cancel the fiber that runs the query, but not the operation itself.

Yes, it is. But cancellation handler that sends the cancel query to the DB is on the fiber that can be cancelled now. So, the blocked fiber will also be canceled from the JDBC side (the database stops the JDBC thread).

> There can be a potential fiber leak with this code

I think there is no fiber leak, the asyncM.blocking fiber will continue running until the JDBC thread finishes dealing with the query. When the JDBC query is completed and the thread is unblocked, that fiber will be cleaned up. Currently, Doobie works in exactly the same way. However, there is a problem with this approach when it comes to background tasks. The cancellation of a fiber also tries to cancel the jdbcBlockedFiber, but the issue is that the cancellation of jdbcBlockedFiber is blocking and doesn't work effectively. So, using background tasks doesn't solve the problem

> What is it attempted to accomplish there exactly?

You cannot cancel jdbcBlockedFiber because it is running on a blocked JDBC thread. However, you can cancel the jdbcBlockedFiber.join.start fiber, because it is blocked asynchronously. This allows the cancellation handler, which sends a cancel query to the database, to be executed. Please refer to the example code in the description, as it illustrates the problem.

Reply:
> I think there is no fiber leak, the asyncM.blocking fiber will continue running until the JDBC thread finishes dealing with the query. When the JDBC query is completed and the thread is unblocked, that fiber will be cleaned up.

Actually, this is what I would call a "fiber leak". Because the fiber goes out of control at this point and lives solely on the mercy of JDBC.

> This allows the cancellation handler, which sends a cancel query to the database, to be executed.

Could you elaborate a bit – why do you think a cancel query is sent to the database in this case? Where is it happening exactly?

I see that you are trying to reproduce it with the test. However, I'm not sure it is working because the query gets cancelled gracefully. Perhaps it is happening just because the connection to the DB gets closed after the cancellation, which in turn cancels the query. However, it may not always work with connection pooling, for example.

> There is a problem with this approach when it comes to background tasks. The cancellation of a fiber also tries to cancel the jdbcBlockedFiber, but the issue is that the cancellation of jdbcBlockedFiber is blocking and doesn't work.

Yes, exactly. Because it guarantees that there's no leak in that case. I feel it was designed that way for a good reason.

Reply:
> Actually, this is what I would call a "fiber leak". Because the fiber goes out of control at this point and lives solely on the mercy of JDBC.

There is no way in CE3 to cancel a fiber running on a blocked thread, I suppose. So the fiber lives solely on the mercy of JDBC anyway.

Reply:
> Could you elaborate a bit – why do you think a cancel query is sent to the database in this case? Where is it happening exactly?

I added a debug log to PreparedStatementInterpreter.close to check if the raw JDBC method (PreparedStatement.close) was called after the fiber was canceled.

Reply:
> There is no way in CE3 to cancel a fiber running on a blocked thread, I suppose. So the fiber lives solely on the mercy of JDBC anyway.

Actually, there are two ways in general: [interruptible](https://github.com/typelevel/cats-effect/blob/be1b65687215b1332a7be80cd6759d147680c04f/kernel/shared/src/main/scala/cats/effect/kernel/Sync.scala#L93) and [cancellable](https://github.com/typelevel/cats-effect/blob/be1b65687215b1332a7be80cd6759d147680c04f/kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala#L264). The former is unlikely to help in this case, but the latter could come in handy, perhaps.

Reply:
See also https://github.com/tpolecat/doobie/issues/1922#issuecomment-1751461106

Reply:
There is one disadvantage with using *cancellable*. While it allows a cancellation handler to be executed even if a fiber is running on a blocking thread, it doesn't work properly when we wrap the call in a timeout. In this scenario, the client’s query to the service will remain stuck until the JDBC thread is unblocked. Below is an illustrative example of this issue in code:

```scala
import cats.effect.{Async, IO, IOApp}

object OnCancelTest extends IOApp.Simple {
  import scala.concurrent.duration._

  def goDB: IO[Unit] = IO(println("START goDB")).map(
    _ => Thread.sleep(10000)
  )
  .guarantee(
    IO(println("SEND CANCELLATION VIA JDBC"))
  )

  override def run: IO[Unit] =
    for {
      f <- goDB.timeout(
        1.seconds
      ).start
      _ <- f.join
      _ <- IO(println("Timeout")) // This prints only after 10 seconds
    } yield ()
}
```

Reply:
So now we have two possible solutions:

1) Don't control the JDBC fiber lifecycle at all: This approach works well with timeouts.

2) Control the JDBC fiber lifecycle using Cancellable: However, this approach leads to issues where timeouts do not work as expected.

Reply:
There is also another workaround: you can run the JDBC query directly on a blocking ExecutionContext using Future, for example. Then, convert it to IO using IO.fromFuture. With this approach, there won't be a JDBC fiber, so there's no need to worry about controlling the fiber's lifecycle. (IO.fromFuture will cancel the fiber instantly, as it is implemented through IO.async).

Reply:
@jatcwang I’d like to hear your opinion as well.  I'd like to understand which approach is more correct.

Reply:
> it doesn't work properly when we wrap the call in a timeout.

@TalkingFoxMid Thanks for raising this, would you be able to give a small demonstration of this problem, ideally using only Cats Effect? I worked with Daniel on `cancelable` in https://github.com/typelevel/cats-effect/pull/3460 and I would like to understand this limitation.

Reply:
@TalkingFoxMid , I'm not sure how you came to the conclusion that `cancelable` does not respect timeouts. Besides, the code that you suggest as an example of the issue doesn't seem having any business with `cancelable` per se (or am I missing some point here?)

On the other hand, I tried it myself and it seems working pretty well to me:
```scala
import cats.effect.*
import scala.concurrent.duration.*

object TimeoutCancellableApp extends IOApp.Simple:
  @volatile
  private var taskCancelled: Boolean = false

  private def blockingButCancellableTask: IO[Unit] = IO.blocking {
    // Runs forever unless canceled.
    while (!taskCancelled) Thread.sleep(100)
  }

  private def cancelTask: IO[Unit] = IO { taskCancelled = true }

  override def run: IO[Unit] =
    blockingButCancellableTask
      .cancelable(cancelTask)
      .timeout(3.seconds)
```
When I run this example, I get the following output:
```
[error] java.util.concurrent.TimeoutException: 3 seconds
```
i.e. it breaks the execution after 3 seconds as requested.

Reply:
@TalkingFoxMid I have pushed a test case using pg_sleep and connection pool.
From what I can tell the cancellation is happening but that's from PreparedStatment#close() being called (due to various `bracket`). Whether the DB-side query is cancelled when PreparedStatment#close() is called is implementation dependent but since pgjdbc does cancel the query I am happy with the general approach.

In terms of the implementation details of the cancellation I'll need to study it more and will be following this discussion closely. Thanks!

Reply:
@TalkingFoxMid @jatcwang ,
Although I'm still not pretty sure it is the best approach with firing a fiber and then relying on some external cancelation magic, but it looks working indeed (I checked) and also it looks simpler to accomplish comparing to a possible version with `cancelable` and explicit cancelation effect.

However, I would probably suggest to add some protection against fiber leaks and also simplify results processing a little, i.e. the `primitive` function might look like the following:
```scala
  def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli { a =>

    asyncM.uncancelable { unmask =>
      for {
        jdbcBlockedFiber <- asyncM.blocking(f(a)).attempt.start
        a <- unmask(jdbcBlockedFiber.join).flatMap(_.embedNever.rethrow)
      } yield a
    }
  }
```
Here, cancelation is masked for the entire execution and fiber creation workout, but then gets unmasked just for the fiber's `join` operation. It should guarantee that there will be no leaked fibers possible. I checked it with both `QueryCancellationSuite` and `HikariQueryCancellationSuite` – both seem working correctly to me.

Reply:
> @TalkingFoxMid , I'm not sure how you came to the conclusion that `cancelable` does not respect timeouts. Besides, the code that you suggest as an example of the issue doesn't seem having any business with `cancelable` per se (or am I missing some point here?)
> 
> On the other hand, I tried it myself and it seems working pretty well to me:
> 
> ```scala
> import cats.effect.*
> import scala.concurrent.duration.*
> 
> object TimeoutCancellableApp extends IOApp.Simple:
>   @volatile
>   private var taskCancelled: Boolean = false
> 
>   private def blockingButCancellableTask: IO[Unit] = IO.blocking {
>     // Runs forever unless canceled.
>     while (!taskCancelled) Thread.sleep(100)
>   }
> 
>   private def cancelTask: IO[Unit] = IO { taskCancelled = true }
> 
>   override def run: IO[Unit] =
>     blockingButCancellableTask
>       .cancelable(cancelTask)
>       .timeout(3.seconds)
> ```
> 
> When I run this example, I get the following output:
> 
> ```
> [error] java.util.concurrent.TimeoutException: 3 seconds
> ```
> 
> i.e. it breaks the execution after 3 seconds as requested.

Consider the following scenario: 

A query is canceled due to a timeout -> 
Triggers a call to PreparedStatement.close() -> 
Cancellation request is sent to the DB ->
However, the DB does not immediately respond to the PreparedStatement.close() request (possibly due to a failover or another issue). In this situation, the application is left waiting for the JDBC operation to complete (because we are trying to guarantee JDBC-fiber closing), causing the user query to hang (and there timeout does not work as expected)

It is not possible to instantly cancel the JDBC operation **via an external state** using cancellable (like it is in your example) in this case. I suppose that calling PreparedStatement.cancel() is the only way to attempt this, but it appears to be unreliable because it depends on the DB availability

Reply:
> > it doesn't work properly when we wrap the call in a timeout.
> 
> @TalkingFoxMid Thanks for raising this, would you be able to give a small demonstration of this problem, ideally using only Cats Effect? I worked with Daniel on `cancelable` in [typelevel/cats-effect#3460](https://github.com/typelevel/cats-effect/pull/3460) and I would like to understand this limitation.


**cancellable** provides a way to cancel a fiber running on a blocked thread by changing an external state, allowing the blocked thread to detect the change and stop execution. However, in our case, there is a potential issue: changing the external state via PreparedStatement.close() may not always be available, as it depends on the database's availability. If the database is slow to respond, the close operation might not execute instantly. This delay occurs because the cancellation request has to be sent to the database over the network, which can introduce additional latency. Consequently, the timeout might not cancel the query after the intended 300ms, but instead after 600ms or more, due to this added latency

This is the only limitation of the Cancelable interface that I see in this context

Reply:
> @TalkingFoxMid @jatcwang , Although I'm still not pretty sure it is the best approach with firing a fiber and then relying on some external cancelation magic, but it looks working indeed (I checked) and also it looks simpler to accomplish comparing to a possible version with `cancelable` and explicit cancelation effect.
> 
> However, I would probably suggest to add some protection against fiber leaks and also simplify results processing a little, i.e. the `primitive` function might look like the following:
> 
> ```scala
>   def primitive[J, A](f: J => A): Kleisli[M, J, A] = Kleisli { a =>
> 
>     asyncM.uncancelable { unmask =>
>       for {
>         jdbcBlockedFiber <- asyncM.blocking(f(a)).attempt.start
>         a <- unmask(jdbcBlockedFiber.join).flatMap(_.embedNever.rethrow)
>       } yield a
>     }
>   }
> ```
> 
> Here, cancelation is masked for the entire execution and fiber creation workout, but then gets unmasked just for the fiber's `join` operation. It should guarantee that there will be no leaked fibers possible. I checked it with both `QueryCancellationSuite` and `HikariQueryCancellationSuite` – both seem working correctly to me.

Could you elaborate on where jdbcBlockedFiber cancellation occurs? I wrote a simple code example, and it doesn't seem that this fiber is being canceled:

```
object OnCancelTest extends IOApp.Simple {
  import scala.concurrent.duration._

  def goDB: IO[Unit] =
    IO.uncancelable(unmask =>
      for {
        jdbcBlockedFiber <- (IO.sleep(10.seconds).onCancel(IO(println("cancel blocked fiber")))).start
        _ <-  unmask(jdbcBlockedFiber.join)
      } yield ()
    )


  override def run: IO[Unit] =
    for {
      f <- goDB.onCancel(IO(println("OnCancel"))).start
      _ <- IO.sleep(1.second)
      _ <- f.cancel
      _ <- IO(println("Done"))
      _ <- IO.never[Unit]
    } yield ()
}
```

"cancel blocked fiber" prints only after 10 seconds, but not after f.cancel is called

Reply:
@satorg  @jatcwang I had prepared an alternative solution there: https://github.com/tpolecat/doobie/pull/2079 
It is based on adding **cancelable** operation to WeakASync

Reply:
@satorg I thought about fiber leak and have an insight. Doesn't fiber leak in that context equals thread leak ? Code is:

```
asyncM.blocking(f(a)).start
```

I suppose there is only one case when fiber can hang forever: if blocking thread hangs too with that task **f(a)** forever.  I don't really think it is relevant to worry about fiber leak here because: fiber leak => thread leak, and we now we allow ourselves to believe that there is no thread leak

Reply:
> If the database is slow to respond, the close operation might not execute instantly. The delay occurs also because the cancellation request has to be sent to the database over the network, which can introduce additional latency. Consequently, the timeout might not cancel the query after the intended 300ms, but instead after 600ms or more, due to this added latency (and it leads to user request stucks)

@TalkingFoxMid Thanks for explaining. If I understood correctly, what you are describing is not a limitation of `cancelable` but actually a general feature of Cats Effect, which is that it backpressures on cancelation i.e. if you request cancelation of some fiber, it will wait until that cancelation fully completes.

Consider this example:
```scala
IO.sleep(2.seconds).onCancel(IO.sleep(1.second)).timeOut(1.second)
```

Even though the timeout is after 1 second, it will take 2 seconds to complete because of the `onCancel` finalizer. Is this what you are considering to be a limitation?

Reply:
> > If the database is slow to respond, the close operation might not execute instantly. The delay occurs also because the cancellation request has to be sent to the database over the network, which can introduce additional latency. Consequently, the timeout might not cancel the query after the intended 300ms, but instead after 600ms or more, due to this added latency (and it leads to user request stucks)
> 
> @TalkingFoxMid Thanks for explaining. If I understood correctly, what you are describing is not a limitation of `cancelable` but actually a general feature of Cats Effect, which is that it backpressures on cancelation i.e. if you request cancelation of some fiber, it will wait until that cancelation fully completes.
> 
> Consider this example:
> 
> ```scala
> IO.sleep(2.seconds).onCancel(IO.sleep(1.second)).timeout(1.second)
> ```
> 
> Even though the timeout is after 1 second, it will take 2 seconds to complete because of the `onCancel` finalizer. Is this what you are considering to be a limitation?


Consider also this example:

```
import cats.effect.{Async, IO, IOApp}
import scala.concurrent.duration._

object OnCancelTest extends IOApp.Simple {
  @volatile var blocked = true

  def goDB: IO[Unit] =
    IO {
      while (blocked) {Thread.sleep(100)}
    }.cancelable(
      {IO.sleep(5.seconds) >> IO {blocked = false}}.start.void
    )

  override def run: IO[Unit] =
    for {
      fiber <- goDB.start
      _ <- IO.sleep(100.millis)
      _ <- fiber.cancel
      _ <- IO(println("Cancelled"))
    } yield ()
}
```

In that snippet, fiber.cancel was blocked not only because it waits for the cancelable handler to complete (the handler completes instantly because it only starts another fiber). The problem also lies in the fact that the external state wasn't modified instantly, even if the cancellation handler completed successfully. This is similar to what we see in Doobie: even after a cancel query is sent to the database, there is still some latency before the database unblocks the blocked JDBC thread running query.

I don't think this is a limitation of CE3, but it does seem like a slight disadvantage of this approach in the context of the current problem in Doobie. 

This approach can cause the user query to hang for a long time because we are trying to wait for the query to be canceled in the database. However, there are situations when you just need to **timeout** immediately, return the default value and continue executing the request further. This is exactly our case, so this disadvantage is relevant for us

Reply:
@satorg Could you give an approve here?

Reply:
@TalkingFoxMid thanks for including that second example, that was very helpful.

> However, there are situations when you just need to **timeout** immediately, return the default value and continue executing the request further. This is exactly our case, so this disadvantage is relevant for us

Cats Effect specifically has a [`timeoutAndForget`](https://github.com/typelevel/cats-effect/blob/caee0e86f815b641af61f19becd4eac27a769f11/core/shared/src/main/scala/cats/effect/IO.scala#L832-L851) API for this.

I agree that sometimes behavior like this is necessary, but that should be the (careful) choice of the user. Once backpressure is removed, it's not possible to restore it. I am dubious about removing backpressure signals in a library, since then users no longer have agency to make the appropriate decision for their specific usecase.

Reply:
@TalkingFoxMid , 
> However, there are situations when you just need to timeout immediately, return the default value and continue executing the request further.

The catch here is that most likely there will be no immediate relief whatsoever. As @jatcwang found out recently, even if we give up here immediately, the query will be canceled as a part of bracket's release effect at the end of the transaction. Therefore in a case of long-running cancel query, the entire transaction will have to await until the cancel query completes anyway.